src/Microsoft.Health.Fhir.Shared.Api/Modules/ModuleRegistrationExtensions.cs

```diff
@@ -0,0 +1,70 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Health.Extensions.DependencyInjection;
+using Microsoft.Health.Fhir.Api.Configs;
+using Microsoft.Health.Fhir.Api.Modules.FeatureFlags.HtmlUi;
+using Microsoft.Health.Fhir.Api.Modules.FeatureFlags.Validate;
+using Microsoft.Health.Fhir.Api.Modules.FeatureFlags.XmlFormatter;
+
+namespace Microsoft.Health.Fhir.Api.Modules
+{
+    public static class ModuleRegistrationExtensions
+    {
+        public static IServiceCollection RegisterModules(
+            this IServiceCollection services,
+            FhirServerConfiguration configuration,
+            KnownModuleNames modules = KnownModuleNames.All)
+        {
+            if ((modules & KnownModuleNames.Anonymization) == KnownModuleNames.Anonymization)
+            {
+#if R4 && !R4B
+                services.RegisterModule<AnonymizationModule>(configuration);
+#endif
+            }
+
+            if ((modules & KnownModuleNames.Fhir) == KnownModuleNames.Fhir)
+            {
+                services.RegisterModule<FhirModule>(configuration);
+            }
+
+            if ((modules & KnownModuleNames.Mediation) == KnownModuleNames.Mediation)
+            {
+                services.RegisterModule<MediationModule>(configuration);
+            }
+
+            if ((modules & KnownModuleNames.Mvc) == KnownModuleNames.Mvc)
+            {
+                services.RegisterModule<MvcModule>(configuration);
+                services.RegisterModule<HtmlUiFeatureModule>(configuration);
+                services.RegisterModule<ValidateFeatureModule>(configuration);
+                services.RegisterModule<XmlFormatterFeatureModule>(configuration);
+            }
+
+            if ((modules & KnownModuleNames.Operations) == KnownModuleNames.Operations)
+            {
+                services.RegisterModule<OperationsModule>(configuration);
+            }
+
+            if ((modules & KnownModuleNames.Persistence) == KnownModuleNames.Persistence)
+            {
+                services.RegisterModule<PersistenceModule>(configuration);
+            }
+
+            if ((modules & KnownModuleNames.Search) == KnownModuleNames.Search)
+            {
+                services.RegisterModule<SearchModule>(configuration);
+            }
+
+            if ((modules & KnownModuleNames.Validation) == KnownModuleNames.Validation)
+            {
+                services.RegisterModule<OperationsModule>(configuration);
+            }
```

Comment:
A caution with this approach: I don't mind the idea of allowing this to be configurable for http or background scenarios, but this is mixing a component built with "scanning conventions" as the method of discovery with explicit registrations. These two assumptions are conflicting and may cause problems with maintenance (i.e. adding a module and not knowing why it isn't working) in future.

Reply:
Thank you, @brendankowitz. I will abandon the PR for now. I tried registering one module at a time and noticed that I ended up registering all the modules except 2. It is because these modules assume their dependencies be registered already by someone else. I will come back to this later since it will require more time to fix the issue.